reps/2023-10-13-accelerator-support.md

```diff
@@ -0,0 +1,399 @@
+@ray.remote(num_gpus=1)
+def task():
+    ...
+
+@ray.remote(num_gpus=1, accelerator_type="A100")
+def task():
+    ...
+
+@ray.remote(num_gpus=1, accelerator_type="Intel-Max-1550")
+def task():
+    ...
+
+@ray.remote(resources={"TPU": 1})
+def task():
+    ...
+
+@ray.remote(resources={"TPU": 1}, accelerator_type="TPU-V4")
+def task():
+    ...
+```
+
+##### Alternatives
+Other alternative APIs considered:
+
+Similar to `num_gpus`, having a top level parameter for each accelerator family.
+One issue is that this will add many top level parameters to `.remote()` that users
+need to learn even though they don't use those accelerators.
+
+```python
+@ray.remote(num_tpus=1)
+def task():
+    ...
+
+@ray.remote(num_tpus=1, accelerator_type="TPU-V4")
+def task():
+    ...
+```
+
+A single `ACCELERATOR` resource for the ability to schedule a task on any accelerator
+and an `accelerator_family` is also introduced to allow specifying the family of the accelerators to use.
+People feel this is less obvious or expclit than having a resource name per accelerator family
+and the ability to schedule a task on any accelerator is not that needed.
+
+```python
+@ray.remote(resources={"ACCELERATOR": 1})
+def task()
+    ...
+
+@ray.remote(resources={"ACCELERATOR": 1}, accelerator_family="TPU")
+def task()
+    ...
+
+@ray.remote(resources={"ACCELERATOR": 1}, accelerator_type="TPU-V4")
+def task()
+    ...
+```
+
+Similar to the above but having a top level `num_accelerators` parameter:
+
+```python
+@ray.remote(num_accelerators=1)
+def task()
+    ...
+
+@ray.remote(num_accelerators=1, accelerator_family="TPU")
+def task()
+    ...
+
+@ray.remote(num_accelerators=1, accelerator_type="TPU-V4")
+def task()
+    ...
+```
+
+##### Runtime Context API
+
+Currently we have `ray.get_gpu_ids()` and we need similar APIs for getting available ids of other accelerators.
+
+The new API will be inside RuntimeContext since it returns accelerator ids that are available to a worker.
+
+```python
+class RuntimeContext:
+    def get_accelerator_ids() -> Dict[str, List[str]]:
+        """
+        Get the current worker's available accelerator ids.
+
+        Returns:
+            A dictionary keyed by the resource name. The values are list
+            of ids. E.g., `{'GPU': ['0', '1']}`, `{'neuron_cores': ['0', '3']}`.
+        """
+```
+
+#### Implementation
+
+Core defines an `AcceleratorManager` base class and each accelerator needs to implement a subclass (e.g. `TPUAcceleratorManager`):
+
+```python
+class AcceleratorManager(ABC):
+  """This class contains all the functions needed for supporting
+  an accelerator family in Ray."""
+
+  @staticmethod
+  @abstractmethod
+  def get_resource_name() -> str:
+      """Get the name of the resource representing this accelerator family.
+
+      Returns:
+          The resource name: e.g., the resource name for Nvidia GPUs is "GPU"
+      """
+
+  @staticmethod
+  @abstractmethod
+  def get_visible_accelerator_ids_env_var() -> str:
```

Comment:
Do we still need this given that get_current_process_visible_accelerator_ids and set_current_process_visible_accelerator_ids are available?

Reply:
Currently it's only used in some error messages

```
raise ValueError(
                    f"Attempting to start raylet with {num_accelerators} "
                    f"{accelerator_resource_name}, "
                    f"but {accelerator_manager.get_visible_accelerator_ids_env_var()} "
                    f"contains {visible_accelerator_ids}."
                )
```